File: 697_Degree_of_an_Array.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/degree-of-an-array/
# Given a non-empty array of non-negative integers nums, the degree of this array is defined as the maximum frequency
# of any one of its elements.
# Your task is to find the smallest possible length of a (contiguous) subarray of nums that has the same degree as nums.

# Count the frequency of each num and record its first and last index. Then find the num or nums with the largest count.
# Finally, find the shortest distance between first and last indices of each maximum count num.
# Time - O(n)
# Space - O(n)

from collections import defaultdict

class Solution(object):
    def findShortestSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        counts, limits = defaultdict(int), {}

        for i, num in enumerate(nums):

            counts[num] += 1

            if num not in limits:
                limits[num] = [i, i]    # first and last indices are both i
            else:
                limits[num][-1] = i     # new last index of i

        max_count, max_nums = 0, []     # list of nums with max_count
        for num, count in counts.items():

            if count == max_count:
                max_nums.append(num)    # new num with same max_count
            elif count > max_count:
                max_nums = [num]        # new unique max
                max_count = count

        shortest = float("inf")
        for num in max_nums:
            shortest = min(shortest, limits[num][1] - limits[num][0] + 1)

        return shortest

File: java/697_Degree_of_an_Array.java
/*
https://leetcode.com/problems/degree-of-an-array/
Given a non-empty array of non-negative integers nums, the degree of this array is defined as the maximum frequency
of any one of its elements.
Your task is to find the smallest possible length of a (contiguous) subarray of nums, that has the same degree as nums.

Use 3 hashmaps to store the count, first and last indices of each num. Then find the most frequent num, breaking ties
by shortest distance between first and last.
Time - O(n)
Space - O(n)
*/

class Solution {
    public int findShortestSubArray(int[] nums) {

        HashMap<Integer, Integer> counts = new HashMap<>();     // map num to its count
        HashMap<Integer, Integer> first = new HashMap<>();      // map num to its first index
        HashMap<Integer, Integer> last = new HashMap<>();       // map num to its last index

        for (int i = 0; i < nums.length; ++i) {

            int count = counts.getOrDefault(nums[i], 0);
            if (count == 0)
                first.put(nums[i], i);

            last.put(nums[i], i);
            counts.put(nums[i], count + 1);
        }

        int shortest = nums.length;
        int mostFrequent = 0;

        for (Map.Entry<Integer, Integer> entry : counts.entrySet()) {

            int num = entry.getKey();
            int count = entry.getValue();
            int distance = last.get(num) - first.get(num) + 1;

            if (count > mostFrequent || count == mostFrequent && distance < shortest) {
                shortest = distance;
                mostFrequent = count;
            }

        }

        return shortest;
    }
}
